cardano-node: 1.21.1 -> 1.22.1

rvl opened the pull request:
### Overview

- Use cardano-node 1.21.2 stack snapshot from input-output-hk/cardano-haskell#33
- Use cardano-node 1.22.1 stack snapshot from input-output-hk/cardano-haskell#34
- Multiple build and test fixes.
- Use a Cardano.Api.Shelley function for metadata type conversion.

### Comments

- Merge input-output-hk/cardano-haskell#34 first.
- Also depends on input-output-hk/cardano-transactions#24 - build fixes required in cardano-transactions package.

rvl committed:
Update version matrix in readme

rvl committed:
Bring back cardano-api haddock in nix-shell

rvl committed:
Update resolver in stack.yaml

rvl committed:
Update for NodeToClient DictVersion constructor

rvl committed:
Replace deprecated symbol

rvl committed:
Reduce cardano-node key file permissions for tests

The node is now refusing to start with other-user-readable key files.

rvl committed:
Set umask for integration tests

So that cardano-node does not complain about file permissions.

rvl committed:
Another cardano-cli option changed

rvl committed:
Also give shelley-test-cluster the umask treatment

rvl committed:
Use Cardano.Api.Shelley.fromShelleyMetaData instead of own

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -626,14 +628,20 @@ withBFTNode tr baseDir params action =
+                let dst = dir </> f
+                copyFile (source </> f) dst
+                restrictFileMode dst
+                pure dst
```
:+1:

rvl committed:
cardano-node: 1.21.2 -> 1.22.1

rvl committed:
Regenerate nix

rvl committed:
Bump cardano-transactions to latest master branch

Anviking committed:
Fix import of Query

rvl committed:
Update network types to include point

rvl committed:
Era types changes

rvl committed:
NodeToClient codecs need a version now

Remove dead code.

rvl committed:
SL.Shelley -> SL.ShelleyEra

rvl committed:
Fix types which became ambiguous

rvl committed:
NodeToClient codecs need a version now

rvl committed:
PoolParams: poolPubKey -> poolId

Anviking committed:
DictVersion -> Version

Anviking committed:
CardanoCodecConfig requres more arguments now

Anviking committed:
Fix two type annotations

Anviking committed:
Seems we don't need vDict anymore

Anviking committed:
Use castHash for EraIndependentTxBody

Anviking committed:
Fix mainnetVersion

rvl committed:
Add constructor patterns for Mary and Allegra blocks

Anviking committed:
Fixup warnings

rvl committed:
Update Shelley.CompatibilitySpec

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -81,7 +81,7 @@ import Network.TypedProtocol.Pipelined
 import Numeric.Natural
     ( Natural )
-import Ouroboros.Consensus.Ledger.Abstract
+import Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
```
Seems like it changes location at every release, hide and seek 2.0 :upside_down_face:

rvl replied:
Yes. As far as I can tell, HFC is not yet covered by `Cardano.Api`.

KtorZ commented on lib/launcher/src/Cardano/Startup/POSIX.hs:
```diff
@@ -29,3 +35,14 @@ installSignalHandlers notify = void $
+-- | Changes permissions of an existing file so that only the owner can read
+-- them.
+restrictFileMode :: FilePath -> IO ()
+restrictFileMode f = setFileMode f ownerReadMode
```
:+1:

KtorZ commented on lib/shelley/cardano-wallet.cabal:
```diff
@@ -43,6 +43,7 @@ library
     , cardano-crypto
     , cardano-crypto-class
     , cardano-crypto-wrapper
+    , cardano-wallet-launcher
```
That dependency seems to be in the wrong direction :thinking: ?

rvl replied:
It's just that the package name is misleading. We can fix that later - it's slightly easier to do without cardano-wallet-jormungandr.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -365,6 +378,10 @@ toCardanoBlockHeader gp = \case
+    BlockMary _ ->
+        error "TODO: unimplemented mary era"
+    BlockAllegra _ ->
+        error "TODO: unimplemented allegra era"
```
So it begins.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -609,7 +609,11 @@ codecVersion = verMap ! nodeToClientVersion
+    (byronCodecConfig sp)
+    ShelleyCodecConfig
+    ShelleyCodecConfig
+    ShelleyCodecConfig
```
Are those two additional other Codecs supposed to be for Mary and Allegra?

rvl replied:
I don't think so - not at the moment. There is this:
```
data instance BlockConfig (ShelleyBlock era) = ShelleyConfig {..}
```
It fits with AllegraEra and MaryEra.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -718,7 +719,9 @@ walletSyncProgress ctx w = do
 
     ti :: TimeInterpreter IO
-    ti = timeInterpreter (ctx ^. networkLayer @t)
+    ti = neverFails
```
This always felt a bit wrong — to motivate the `neverFails` _outside_ the `syncProgress` function, with `syncProgress` internals as arguments — but I wasn't sure how to fix.

We could make `syncProgress` take a `TimeInterpreter ExcepT PastHorizonException IO` and return `IO SyncProgress`, but then we couldn't use it with `TimeInterpreter Identity` …

_Maybe this still can sort-of work out though... if we hoist the `TimeInterpreter Identity` into `TimeInterpreter PastHorizonException Identity` when needed on the call site_

KtorZ replied:
Although `walletSyncProgress` "only converts times at the tip", there's no guarantee that the tip itself isn't on an unsafe zone, isn't it?

Anviking replied:
The safe-zone is at least a significant portion of an epoch. And we expect the wallet tip to be behind the node tip.

Anviking committed:
Bump ouroboros-consensus to include unsafeExtendSafeZone

Anviking committed:
Re-write TimeInterpreter and use unsafeExtendSafeZone where needed

TimeInterpreter was re-written to have more powerful combinators and
useful logging.

To fix integration tests, unsafeExtendSafeZone was used in the following
places:

1. getTxExpiry for setting TTL in the future
2. mkApiTransaction for showing TTL in the future
3. mkShelleyWallet for delegation status changes
4. combineDbAndLsqData for future pool retirement dates

This allows dates, slots and epochs to be converted between even in the
Shelley era of Byron;Shelley;Allegra;Mary — but the results could also
be **wrong**.

It seems the slotting parameters won't change in the coming eras. It
would still be nice to replace unsafeExtendSafeZone with something else,
in the future.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1640,29 +1643,25 @@ signPayment ctx wid argGenChange mkRewardAccount pwd md ttl cs = db & \DBLayer{.
 getTxExpiry ti maybeTTL = do
     expTime <- addUTCTime ttl <$> getCurrentTime
-    try $ ti $ ceilingSlotAt expTime
+    interpretQuery (unsafeExtendSafeZone ti) (ceilingSlotAt expTime)
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -698,9 +699,25 @@ mkShelleyWallet ctx wid cp meta pending progress = do
+    -- 2. It shouldn't be the end of the world to return the wrong time.
+    --
+    -- But ultimately, we might want to make the uncertainty transparent to API
+    -- users.
```
Perhaps link this comment to a JIRA technical-debt ticket or this will never happen?

Anviking replied:
Done -> ADP-575

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley.hs:
```diff
@@ -333,7 +335,8 @@ serveWallet
                 (Pool.defaultFilePath <$> dir)
-                (timeInterpreter nl)
+                (neverFails "withPoolsMonitoring never forecasts into the future"
+                    $ timeInterpreter nl)
```
What about stake pool retirements :thinking:  ? Or does the conversion only happens with the API-layer's time interpreter?

Anviking replied:
Hm, yeah, not obvious. This is used only for
1. [`selectPoolProduction`](https://github.com/input-output-hk/cardano-wallet/blob/4df2ecc5d5259ce333de6b2dfa39b6ac51b17652/lib/core/src/Cardano/Pool/DB/Sqlite.hs#L898-L908) which was jormungandr-specific and we should remove
2. [rolling back stake distributions](https://github.com/input-output-hk/cardano-wallet/blob/4df2ecc5d5259ce333de6b2dfa39b6ac51b17652/lib/core/src/Cardano/Pool/DB/Sqlite.hs#L487), which I guess also is unused and should be removed

Changing `"withPoolsMonitoring"` in the reason to `"pool db"` might help — if we don't want to clean this up now — but I think a separate PR would be better.

rvl committed:
Separate tracing and result handling for TimeInterpreter

rvl committed:
Fix merge skew
